contracts/0.4.24/Lido.sol

```diff
@@ -645,33 +584,212 @@ contract Lido is ILido, StETH, AragonApp {
+     * @return treasuryFee treasury fee in base precision
+     */
+    function getFeeDistribution() public view returns (uint96 modulesFee, uint96 treasuryFee) {
+        (, uint96[] memory moduleFees, uint96 totalFee, ) = getStakingRouter().getStakingRewardsDistribution();
+        for (uint256 i; i < moduleFees.length; ++i) {
+            modulesFee += moduleFees[i];
+        }
+        treasuryFee = totalFee - modulesFee;
+    }
+
+    /**
+     * @notice Returns current withdrawal credentials of deposited validators
+     * @dev DEPRECATED: use StakingRouter.getWithdrawalCredentials() instead
+     */
+    function getWithdrawalCredentials() public view returns (bytes32) {
+        return getStakingRouter().getWithdrawalCredentials();
+    }
+
+    /**
+     * @notice Returns address of the contract set as LidoExecutionLayerRewardsVault
+     */
     function getELRewardsVault() public view returns (address) {
         return EL_REWARDS_VAULT_POSITION.getStorageAddress();
     }
 
+    /// @dev updates beacon state according to the current report
+    function _processBeaconStateUpdate(
+        uint256 _postBeaconValidators,
+        uint256 _postBeaconBalance
+    ) internal returns (uint256 appearedValidators) {
+        uint256 depositedValidators = DEPOSITED_VALIDATORS_POSITION.getStorageUint256();
+        require(_postBeaconValidators <= depositedValidators, "REPORTED_MORE_DEPOSITED");
+
+        uint256 preBeaconValidators = BEACON_VALIDATORS_POSITION.getStorageUint256();
+        require(_postBeaconValidators >= preBeaconValidators, "REPORTED_LESS_VALIDATORS");
+
+        // Save the current beacon balance and validators to
+        // calculate rewards on the next push
+
+        BEACON_BALANCE_POSITION.setStorageUint256(_postBeaconBalance);
+
+        if (_postBeaconValidators > preBeaconValidators) {
+            BEACON_VALIDATORS_POSITION.setStorageUint256(_postBeaconValidators);
+        }
+
+        return _postBeaconValidators.sub(preBeaconValidators);
+    }
+
+    /// @dev update BufferedEtherReserve to `_newBufferedEtherReserve`
+    function _processBufferedEtherReserveUpdate(uint256 _newBufferedEtherReserve) internal {
+        uint256 minEtherReserve = 0;
+
+        // We are trying to avoid the situation when withdrawal queue is blocked by large request
+        // while ether is constantly spent to deposit new validators
+        // so if we pass a reserve amount less than amount of eth required to finalize the last request
+        // in the queue, we are updating it to that amount
+        address withdrawalQueueAddress = getWithdrawalQueue();
+
+        if (withdrawalQueueAddress != address(0)) {
+            IWithdrawalQueue withdrawalQueue = IWithdrawalQueue(withdrawalQueueAddress);
+
+            uint256 lastUnfinalizedRequestId = withdrawalQueue.finalizedRequestsCounter();
+            (,,minEtherReserve,,,) = withdrawalQueue.getWithdrawalRequestStatus(lastUnfinalizedRequestId);
```

Comment:
Request with id `lastUnfinalizedRequestId` may not exist. It works and `minEtherReserve` should = 0 in this case, but it’s not obvious. If `getWithdrawalRequestStatus` implementation changes, this could lead to errors. I would suggest handling this case more clearly

Reply:
I also do not understand why we block ETH equal to the last request
- we can have more than 1 request
- the last request may not be included in the list of requests that we will want to fulfill If it's close to the next report and block number limit is reached